lib/src/clojure_lsp/feature/file_management.clj

```diff
@@ -249,42 +247,42 @@
                  "Watched files analyzed, took %s"
                  (lsp.kondo/run-kondo-on-paths! filenames db* {:external? false} nil))]
     (swap! db* lsp.kondo/db-with-results result)
-    (f.diagnostic/publish-all-diagnostics! filenames components)
+    (f.diagnostic/publish-all-diagnostics! uris components)
     (producer/refresh-test-tree producer uris)
     (doseq [uri uris]
       (when (get-in @db* [:documents uri :v])
         (when-let [text (shared/slurp-uri uri)]
           (swap! db* assoc-in [:documents uri :text] text))))))
 
-(defn ^:private db-without-file [state-db uri filename]
+(defn ^:private db-without-uri [state-db uri]
   (-> state-db
-      (dep-graph/remove-file uri filename)
+      (dep-graph/remove-doc uri)
       (shared/dissoc-in [:documents uri])
-      (shared/dissoc-in [:analysis filename])
-      (shared/dissoc-in [:findings filename])))
+      (shared/dissoc-in [:analysis uri])
+      (shared/dissoc-in [:findings uri])))
 
-(defn ^:private file-deleted [{:keys [db*], :as components} uri filename]
-  (swap! db* db-without-file uri filename)
-  (f.diagnostic/publish-empty-diagnostics! uri components))
+(defn ^:private files-deleted [{:keys [db*], :as components} uris]
+  (swap! db* #(reduce db-without-uri % uris))
+  (f.diagnostic/publish-empty-diagnostics! uris components))
 
 (defn did-change-watched-files
   [changes
    {:keys [db* watched-files-chan] :as components}]
-  (doseq [{:keys [uri type]} changes]
-    (case type
-      :created (async/>!! watched-files-chan uri)
-      :changed (when (settings/get @db* [:compute-external-file-changes] true)
-                 (async/>!! watched-files-chan uri))
-      :deleted (shared/logging-task
-                 :delete-watched-file
-                 (file-deleted components uri (shared/uri->filename uri))))))
+  (let [{:keys [created changed deleted]} (group-by :type changes)
+        observe-changed? (settings/get @db* [:compute-external-file-changes] true)]
+    (doseq [created-or-changed (concat created (when observe-changed? changed))]
+      (async/>!! watched-files-chan (:uri created-or-changed)))
+    (when (seq deleted)
+      (shared/logging-task
+        :delete-watched-files
+        (files-deleted components (map :uri deleted))))))
```

Comment:
Processing all the deleted URIs at once lets us use `onto-chan!`, which helps avoid "too many pending puts" exceptions.

Reply:
We're already changing the order somewhat by grouping together several changes into one analysis batch. But I think you're right—order doesn't matter much for didChangeWatchedFiles, and it's best to treat everything as one big batch if possible.